### PR TITLE
Improve escaping - Never have end user input crash a LaTeX build.

### PR DESCRIPTION
**Pull request by CarliJoy**
Hi,

I recognized that the escape function did not escape all characters that could lead to a failed LaTeX build on an user input.
Therefore I added some escapes based on https://tex.stackexchange.com/questions/34580/escape-character-in-latex).

Doing so I made sure that also the spacing is kept as an end user would expect it.
See the `/escape` view

The resulting PR also tests this with different latex compilers (I added xelatex) in form of an integration test.
My goal was to make the build as stable as possible, so that an end user won't be able to crash the LaTeX build with inputs.

I achieved the goal so far for *luatex* and *xelatex*. *pdflatex* and *mklatex* still suffer from weak unicode support (see tests).

Be aware that I was so bold and updated the Django version within the PR -> 1.1 is not supported any more. Version 2.2 is the current LTS version.
If you don't like that only reverse back to the old URL schema in `tests/urls.py`.

Note:
Of course the goal of never crashing LaTeX build is only achieved if all inputs are escape manually. To cover this problem I opened an [issue in jinja](https://github.com/pallets/jinja/issues/1377).

### CODE REVIEW

**Comment by CarliJoy** [comment]
Hmm just found: https://github.com/mbr/latex
Maybe a join of both projects is possible?
Or we could use some this project as base for this package?

File: django_tex/filters.py [hunk 1]
```diff
@@ -1,20 +1,46 @@
...
     return (
-        value.replace("&", "\\&")
-        .replace("$", "\\$")
-        .replace("%", "\\%")
-        .replace("#", "\\#")
-        .replace("_", "\\_")
-        .replace("{", "\\{")
-        .replace("}", "\\}")
+        # Replace Backslashes so we dont' double replace backslashes
+        value.replace("\\", BACKSLASH_REPLACEMENT)
```
Comment (weinbusch) [inline django_tex/filters.py:21]: Why replace `\\` with `BACKSLASH_REPLACEMENT`? I would say, if the input string contains a `\\`, the user wants it to be replaced with `\textbackslash\textbackslash` and not with a single `\textbackslash`.

File: README.md [hunk 2]
```diff
@@ -101,7 +101,12 @@ settings. Use the filter like this: `{{ foo|localize }}`.
 If you want to convert linebreaks into LaTeX linebreaks (`\\`), use the `linebreaks` filter (`{{ foo | linebreaks }}`).
 ### Escaping LaTeX special characters
 
-To escape LaTeX special characters, use the `escape_latex` filter. This escapes the following characters: `&$%#_{}`.
+To escape LaTeX special characters, use the `escape_latex` filter, i.e. `{{ foo | escape_latex }}`s.
```
Comment (weinbusch) [inline README.md:104]: Delete the `s` at the end.

File: django_tex/filters.py [hunk 3]
```diff
@@ -1,20 +1,46 @@
...
+        # If there was a space before the character we assume the user wants
+        # it also in the displayed LaTeX text
+        .replace(BACKSLASH_REPLACEMENT + " ", r"\textbackslash\space ")
```
Comment (weinbusch) [inline django_tex/filters.py:24]: Instead of pattern matching for the whitespace and inserting a `\space` command, we should probably use `{}` at the end of the replacement, e.g. `\` gets escaped to `\textbackslash{}`.

File: django_tex/filters.py [hunk 4]
```diff
@@ -1,20 +1,46 @@
...
+        .replace("&", r"\&")
+        .replace("$", r"\$")
+        .replace("%", r"\%")
+        .replace("#", r"\#")
+        .replace("_", r"\_")
+        .replace("{", r"\{")
+        .replace("}", r"\}")
+        # Replace vertical tab as this seems to be the only character from
+        # string.printable that can't be consumed from LaTeX
+        .replace(chr(11), "")
```
Comment (weinbusch) [inline django_tex/filters.py:43]: This might be a matter of taste, but I think, it would be better to let LaTeX throw an error, if the input string contains characters that cannot be consumed, instead of silently removing bad characters.

File: tests/contants.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import string
```
Comment (weinbusch) [inline tests/contants.py:1]: The file name should probably read `constants.py`.

File: tests/tests.py [hunk 6]
```diff
@@ -241,6 +242,33 @@ def test_compile_template_with_graphics_pdflatex(self):
         pdf = compile_template_to_pdf(template_name, context)
         self.assertIsNotNone(pdf)
 
+    def do_compile_escape(self, special_unicode: bool = False):
```
Comment (weinbusch) [inline tests/tests.py:245]: This entire block can be simplified by using the `TestCase.subTest` context manager ([](https://docs.python.org/3/library/unittest.html#subtests) and `itertools.product`. I can submit a pr for that ...

File: tests/tests.py [hunk 7]
```diff
@@ -293,21 +321,40 @@ def test_rendering_unicode(self):
...
-    def test_escaping_special_characters(self):
+    def check_escape(self, value: str, expected: str):
         template_string = "{{ value | latex_escape }}"
-        context = {"value": "&$%#_{}"}
+        context = {"value": value}
         output = self.render_template(template_string, context)
-        expected = "\\&\\$\\%\\#\\_\\{\\}"
         self.assertEqual(output, expected)
 
+    def test_escape_no_space(self):
```
Comment (weinbusch) [inline tests/tests.py:330]: Is there some code repetition here? Why don't we use the unicode characters from the `constants.py` module? Or ditch the `constants.py` module and copy all constants into the `tests.py` module for better readability.

**Review by weinbusch — COMMENTED** [review]
Hi there, thanks a lot for you input. I left some comments for you to consider. I also had some tests failing on my machine and will report on that later.

File: django_tex/filters.py (hunk 1 above)
Comment (CarliJoy) [reply django_tex/filters.py:21]: Simply put `print("\\")` in your python console to se what is happening.

Having "\\" is the only way to define a __single__ backslash.

File: django_tex/filters.py (hunk 3 above)
Comment (CarliJoy) [reply django_tex/filters.py:24]: agreed

File: django_tex/filters.py (hunk 4 above)
Comment (CarliJoy) [reply django_tex/filters.py:43]: The whole idea was that end user (who input the variables) are never able to crash the LaTeX generation, no matter what is inputed.
What is use of the escape otherwise?

File: tests/tests.py (hunk 6 above)
Comment (CarliJoy) [reply tests/tests.py:245]: A PR would be great. Have not much experience with pure unittest -> mostly use pytest atm.

File: tests/contants.py (hunk 5 above)
Comment (CarliJoy) [reply tests/contants.py:1]: I will move it.

File: django_tex/filters.py (hunk 1 above)
Comment (weinbusch) [reply django_tex/filters.py]: Oh, you're right. I was confused by the comment about not double replacing backslashes. I think, the same can be achieved using a regex subsitution as in commit 104163e below.

File: tests/tests.py (hunk 6 above)
Comment (weinbusch) [reply tests/tests.py]: See commits 80baed9 ff. below.

File: django_tex/filters.py (hunk 4 above)
Comment (weinbusch) [reply django_tex/filters.py]: I think, users should not put an illegal character in the input string in the first place. And if they do, they should be notified by an error message. For me, escaping means translating a legal character that has a certain meaning within the context of the prgramming language, i.e. a control character used by latex, into another legal character that has a different meaning.

File: tests/tests.py (hunk 6 above)
Comment (weinbusch) [reply tests/tests.py]: However, ... do we really need to test how different variants of latex engines respond to unicode? The more I think about it, the more I believe that these tests are outside of the scope of `django-tex`. Plus the tests are awfully slow.

Comment (CarliJoy) [reply tests/tests.py]: Being a django app the user input is most likely being inputed by end users.
So actually considering also the security risks that come from crashing an application I would definitely see it in the scope of django-tex. 🤷🏻

Comment (CarliJoy) [reply tests/tests.py]: Another option would be that I release the escape filter itself as separate package with all the test and you just import them.
Doing so your test would only require the django + tex related stuff.

What do you think?

**Comment by weinbusch** [comment]
I did some test cleanup and I will merge this pr.

I did not add a filter for replacing `chr(11)` as you suggested, for the reasons mentioned above. If the user enters illegal charachters, these should either be filtered out by form validation or (better) the user should be notified by an error. See also this piece of haskell magic, where no such replacement is done, at least as far as I can tell: [https://github.com/jgm/pandoc/src/Text/Pandoc/Writers/LaTeX/Util.hs](https://github.com/jgm/pandoc/blob/961268446c551e43b43cd3dffaf0f3c6be4dfa6f/src/Text/Pandoc/Writers/LaTeX/Util.hs#L102-L177)
